### PR TITLE
Fix removal of closure variables in types

### PR DESCRIPTION
**Pull request by lthls**
I plan to add examples that would fail without this patch; please don't merge before I've done that.

### CODE REVIEW

**Comment by lthls** [comment]
It's actually very hard to create an example which would fail without this patch.
The reason is actually tricky (and possibly a bug): when renaming types (in particular, on import) we first check whether the renaming is empty, in which case we return the type as is. But that means that if the type contains closure variables that should be removed, then this removal is not done.
Since for small files the renaming for import is often trivial (no hash conflict), it's very hard to come up with a small example that actually removes closure variables when importing types.

I'll think a bit more about this later, in particular at some point we needed to remove the unused closure entries as we could get unbound code ID errors otherwise; I'm not sure whether this is still the case or not.

I think I'll push an example that didn't actually fail, but would have failed if not for this shortcut, to illustrate this PR.

File: middle_end/flambda2/nominal/renaming.ml [hunk 1]
```diff
@@ -63,18 +64,21 @@ end = struct
...
+         types to remove occurrences of unused closure variables, as otherwise
+         the types can contain references to code that is neither exported nor
+         present in the actual object file. But this means rewriting types, and
+         the only place a rewriting traversal is done at the moment is during
+         import. This solution is not ideal because the missing code IDs will
+         still be present in the emitted cmx files, and during the traversal in
+         [Flambda_cmx.compute_reachable_names_and_code] we have to assume that
+         code IDs can be missing (and so we cannot detect code IDs that are
+         really missing at this point). *)
+      compilation_unit : Compilation_unit.t
```
Comment (chambart) [inline middle_end/flambda2/nominal/renaming.ml:78]: I would also call that one `original_compilation_unit`

**Review by chambart — APPROVED** [review]
This is good. It would be nicer to do that while exporting rather than importing, but this is the simplest fix.

File: middle_end/flambda2/nominal/renaming.ml [hunk 2]
```diff
@@ -138,7 +144,10 @@ end = struct
     | exception Not_found -> simple
 
   let closure_var_is_used t var =
-    Var_within_closure.Set.mem var t.used_closure_vars
+    if Var_within_closure.in_compilation_unit var t.original_compilation_unit
+    then Var_within_closure.Set.mem var t.used_closure_vars
+    else (* This closure variable may be used in other units *)
```
Comment (mshinwell) [inline middle_end/flambda2/nominal/renaming.ml:149]: Please change "may" to "might".

**Review by mshinwell — APPROVED** [review]

**Comment by lthls** [comment]
I've added examples (though as explained above, these don't actually fail currently even though they should).

File: middle_end/flambda2/nominal/renaming.ml (hunk 2 above)
Comment (lthls) [reply middle_end/flambda2/nominal/renaming.ml]: Done

**Review by lukemaurer — APPROVED** [review]

**Pull request merged**